src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,118 @@
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_INDEX = 0;
+    private static final int INITIAL_CAPACITY = 10;
+    private Object[] defaultArray = new Object[INITIAL_CAPACITY];
+    private int sizeCounter = 0;
+
     @Override
     public void add(T value) {
-
+        if (sizeCounter == defaultArray.length) {
+            defaultArray = grow(defaultArray);
+        }
+        defaultArray[sizeCounter++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        rangeCheckerForAdding(index);
+        if (++sizeCounter == defaultArray.length) {
+            defaultArray = grow(defaultArray);
+        }
+        System.arraycopy(defaultArray, index, defaultArray, index + 1, sizeCounter - index);
+        defaultArray[index] = value;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (sizeCounter + list.size() > defaultArray.length) {
+            defaultArray = grow(defaultArray);
+        }
+        increaseArray(list);
+    }
+
+    private void increaseArray(List<T> list) {
+        Object[] tempList = new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            tempList[i] = list.get(i);
+        }
+        System.arraycopy(tempList, DEFAULT_INDEX, defaultArray, sizeCounter, tempList.length);
+        sizeCounter += list.size();
+    }
+
+    public Object[] grow(Object[] array) {
+        int arrayLength = defaultArray.length + (defaultArray.length / 2);
+        Object[] temp = new Object[arrayLength];
+        System.arraycopy(array, DEFAULT_INDEX, temp, DEFAULT_INDEX, sizeCounter);
+        return temp;
+    }
 
+    private void rangeChecker(int index) {
+        if (index < 0 || index >= sizeCounter) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of bounds, for length " + size());
+        }
+    }
+
+    private void rangeCheckerForAdding(int index) {
+        if (index < 0 || index > sizeCounter) {
+            throw new ArrayListIndexOutOfBoundsException("Index: " + index
+                    + " out of bounds, for length " + size());
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        rangeChecker(index);
+        return (T)defaultArray[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        rangeChecker(index);
+        defaultArray[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        rangeChecker(index);
+        T removedElement = (T) defaultArray[index];
+        if (index == sizeCounter - 1) {
+            System.arraycopy(defaultArray, DEFAULT_INDEX, defaultArray,
+                    DEFAULT_INDEX, sizeCounter - 1);
+            sizeCounter--;
+            return removedElement;
+        }
+        System.arraycopy(defaultArray, index + 1, defaultArray, index, sizeCounter - index);
```

Comment:
```suggestion
        System.arraycopy(defaultArray, index + 1, defaultArray, index, sizeCounter - index - 1);
```
with this you can remove if that above this line
try to figure out why

Reply:
because it was "index + 1"  and so length to add can be additionally minus 1 if  i counted right